datanode/datanode.go

```diff
@@ -0,0 +1,105 @@
+package main
+
+import (
+	"log/slog"
+	"math/rand"
+	"net/rpc"
+	"time"
+
+	"github.com/ZhenbangYou/TinyDFS/tiny-dfs/common"
+)
+
+type DataNode struct {
+	nameNodeEndpoint string
+	dataNodeEndpoint string
+}
+
+func (datanode *DataNode) registerWithNameNode() bool {
+	slog.Info("Registering with namenode",
+		"nameNodeEndpoint", datanode.nameNodeEndpoint,
+		"dataNodeEndpoint", datanode.dataNodeEndpoint)
+
+	// retry if failed
+	for {
+		client, err := rpc.DialHTTP("tcp", datanode.nameNodeEndpoint)
+		if err != nil {
+			slog.Error("error dialing namenode", "error", err)
+			// retry after a random interval
+			delay := common.RETRY_MIN_INTERVAL + time.Duration(rand.Int63n(int64(common.RETRY_MAX_INTERVAL-common.RETRY_MIN_INTERVAL)))
+			time.Sleep(delay)
+			continue
+		}
+
+		var success bool
+		err = client.Call("NameNode.RegisterDataNode", datanode.dataNodeEndpoint, &success)
+		if err != nil || !success {
+			slog.Error("error registering with namenode", "error", err)
+			return false
+		}
+		return true
+	}
+}
+
+func (datanode *DataNode) sendBlockReport() bool {
+	client, err := rpc.DialHTTP("tcp", datanode.nameNodeEndpoint)
+	if err != nil {
+		slog.Error("error dialing namenode", "error", err)
+		return false
+	}
+
+	blockReport := datanode.generateBlockReport()
+	var success bool
+	err = client.Call("NameNode.ReceiveBlockReport", blockReport, &success)
+	if err != nil || !success {
+		slog.Error("error sending block report to namenode", "error", err)
+		return false
+	}
+	return true
+}
+
+func (datanode *DataNode) heartbeatLoop() {
```

Comment:
though the function name may seem self-explained, I think there're still details worth commenting:
1. HB is sent periodically; if failing, retry after a certain interval periodically and indefinitely;
2. this function also sends block reports after a failed HB.